core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -408,11 +450,15 @@ private TableMetadata applyChangesToMapping(TableMetadata metadata) {
   private static Schema applyChanges(Schema schema, List<Integer> deletes,
                                      Map<Integer, Types.NestedField> updates,
                                      Multimap<Integer, Types.NestedField> adds,
-                                     Multimap<Integer, Move> moves) {
+                                     Multimap<Integer, Move> moves,
+                                     Set<String> rowIdentifierNames) {
     Types.StructType struct = TypeUtil
         .visit(schema, new ApplyChanges(deletes, updates, adds, moves))
         .asNestedType().asStructType();
-    return new Schema(struct.fields());
+    Set<Integer> refreshedRowIdentifiers = rowIdentifierNames.stream()
+        .map(name -> struct.field(name).fieldId())
+        .collect(Collectors.toSet());
+    return new Schema(struct.fields(), refreshedRowIdentifiers);
```

Comment:
I think we may want a change to `TableMetadata` to ensure an update to only row identifier should still create a new schema; I think currently if a schema update does not change the underlying `struct`, we will consider the operation a no-op (comparison on `struct` is also why we don't have `equals`/`hashcode` in `Schema`)

Reply:
thanks! I completely forgot that part... I added a new method `sameSchema` and now use that to check equality in table metadata.